resolver/parallel_best_resolver.go

```diff
@@ -152,89 +134,149 @@ func (r *ParallelBestResolver) Name() string {
 	logger := log.WithPrefix(request.Log, parallelResolverType)
 
-	var resolvers []*upstreamResolverStatus
-	for _, r := range r.resolversPerClient {
-		resolvers = r
+	if len(r.resolvers) == 1 {
+		logger.WithField("resolver", r.resolvers[0].resolver).Debug("delegating to resolver")
 
-		break
+		return r.resolvers[0].resolver.Resolve(request)
 	}
 
-	if len(resolvers) == 1 {
-		logger.WithField("resolver", resolvers[0].resolver).Debug("delegating to resolver")
+	ctx := context.Background()
+
+	// using context with timeout for random upstream strategy
+	if r.resolverCount == 1 {
+		var cancel context.CancelFunc
 
-		return resolvers[0].resolver.Resolve(request)
+		logger = log.WithPrefix(logger, "random")
+		timeout := config.GetConfig().Upstreams.Timeout
+
+		ctx, cancel = context.WithTimeout(ctx, time.Duration(timeout))
+		defer cancel()
 	}
 
-	r1, r2 := pickRandom(resolvers)
-	logger.Debugf("using %s and %s as resolver", r1.resolver, r2.resolver)
+	resolvers := pickRandom(r.resolvers, r.resolverCount)
+	ch := make(chan requestResponse, len(resolvers))
 
-	ch := make(chan requestResponse, resolverCount)
+	// build usedResolver log string
+	var usedResolvers string
+	for _, resolver := range resolvers {
+		usedResolvers += fmt.Sprintf("%q,", resolver.resolver)
+	}
 
-	var collectedErrors []error
+	usedResolvers = strings.TrimSuffix(usedResolvers, ",")
 
-	logger.WithField("resolver", r1.resolver).Debug("delegating to resolver")
+	logger.Debug("using " + usedResolvers + " as resolver")
 
-	go r1.resolve(request, ch)
+	for _, resolver := range resolvers {
+		logger.WithField("resolver", resolver.resolver).Debug("delegating to resolver")
 
-	logger.WithField("resolver", r2.resolver).Debug("delegating to resolver")
+		go resolver.resolve(request, ch)
+	}
 
-	go r2.resolve(request, ch)
+	response, collectedErrors := evaluateResponses(ctx, logger, ch, resolvers)
+	if response != nil {
+		return response, nil
+	}
+
+	if !r.retryWithDifferentResolver {
+		return nil, fmt.Errorf("resolution was not successful, used resolvers: %s errors: %v",
+			usedResolvers, collectedErrors)
+	}
 
-	for len(collectedErrors) < resolverCount {
-		result := <-ch
+	return r.retryWithDifferent(logger, request, resolvers)
+}
 
-		if result.err != nil {
-			logger.Debug("resolution failed from resolver, cause: ", result.err)
-			collectedErrors = append(collectedErrors, result.err)
-		} else {
-			logger.WithFields(logrus.Fields{
-				"resolver": *result.resolver,
-				"answer":   util.AnswerToString(result.response.Res.Answer),
-			}).Debug("using response from resolver")
+func evaluateResponses(
+	ctx context.Context, logger *logrus.Entry, ch chan requestResponse, resolvers []*upstreamResolverStatus,
+) (*model.Response, []error) {
+	var collectedErrors []error
 
-			return result.response, nil
+	for len(collectedErrors) < len(resolvers) {
+		select {
+		case <-ctx.Done():
+			// this context is currently only set & canceled if resolverCount == 1
```

Comment:
A bit pedantic but canceled and deadline/timeout aren't the same thing for contexts:
```suggestion
			// this context currently only a deadline when resolverCount == 1
```